### PR TITLE
add resolveAddressesWithAtos()

### PR DESCRIPTION
**Pull request by WalterBright**
This comes from https://github.com/ldc-developers/ldc/commit/501c6577fa052c3b8d69373105aec1c07574c6e5

ping @kinke @JohanEngelen

### CODE REVIEW

**Comment by JohanEngelen** [comment]
Nice. I no longer remember why I needed to adjust the test makefile and text output in the LDC commit. Seems it is not needed here.

File: druntime/src/core/internal/backtrace/dwarf.d [hunk 1]
```diff
@@ -263,6 +268,141 @@ int processCallstack(Location[] locations, const(ubyte)[] debugLineSectionData,
     return 0;
 }
 
+version (Darwin) {
```
Comment (dkorpel) [inline druntime/src/core/internal/backtrace/dwarf.d:271]:
```suggestion
version (Darwin) 
{
```

File: druntime/src/core/internal/backtrace/dwarf.d [hunk 2]
```diff
@@ -263,6 +268,141 @@ int processCallstack(Location[] locations, const(ubyte)[] debugLineSectionData,
...
+    {
+        import core.stdc.stdio : fclose, fflush, fgets, fprintf, printf, snprintf;
+        import core.stdc.stdlib : exit;
+        import core.sys.posix.stdio : fdopen;
+        import core.sys.posix.unistd : close, dup2, execlp, fork, getpid, pipe;
+        // Create in/out pipes to communicate with the forked exec
+        int[2] dummy_pipes; // these dummy pipes are there to prevent funny issues when stdin/stdout is closed and pipe returns id 0 or 1
+        int[2] pipes_to_atos;
+        int[2] pipes_from_atos;
+        if ( pipe(dummy_pipes) < 0 || pipe(pipes_to_atos) < 0 || pipe(pipes_from_atos) < 0 ) {
```
Comment (dkorpel) [inline druntime/src/core/internal/backtrace/dwarf.d:295]:
```suggestion
        if ( pipe(dummy_pipes) < 0 || pipe(pipes_to_atos) < 0 || pipe(pipes_from_atos) < 0 ) 
        {
```

**Review by dkorpel — APPROVED** [review]

**Comment by JohanEngelen** [comment]
> Nice. I no longer remember why I needed to adjust the test makefile and text output in the LDC commit. Seems it is not needed here.

In the LDC PR I read the reason: it is to be independent of build folder of dmd. If the developer is using a long-ish path name for DMD development, the current test will truncate at different locations and fail.

**Comment by WalterBright** [comment]
@JohanEngelen should this be merged into dmd or not?

**Comment by JohanEngelen** [comment]
> @JohanEngelen should this be merged into dmd or not?

Yes, great.
But it'd be good to also add the testcase changes from the LDC commit. It prevents developer troubles.

**Comment by WalterBright** [comment]
Will it resolve the complaint that dmd suddenly stopped working on OSX?

**Comment by JohanEngelen** [comment]
> Will it resolve the complaint that dmd suddenly stopped working on OSX?

I don't know what "stopped working" means, sorry. This change fixes the lack of debug file:line info on macOS.

**Comment by WalterBright** [comment]
> @JohanEngelen I don't know what "stopped working" means, sorry.

https://www.digitalmars.com/d/archives/digitalmars/D/D_has_become_unbearable_and_it_needs_to_stop_369162.html#N369573

**Comment by JohanEngelen** [comment]
> > @JohanEngelen I don't know what "stopped working" means, sorry.
> 
> https://www.digitalmars.com/d/archives/digitalmars/D/D_has_become_unbearable_and_it_needs_to_stop_369162.html#N369573

Ah OK. The issue is here: https://forum.dlang.org/post/himxshnmnapmyipsvqlq@forum.dlang.org
The fix is described here: https://github.com/ldc-developers/ldc/issues/3864#issuecomment-1229813021

The `atos` druntime code is there to support file:line info in backtraces when the debug section is stripped from the executable.

File: druntime/src/core/internal/backtrace/dwarf.d (hunk 2 above)
Comment (WalterBright) [reply druntime/src/core/internal/backtrace/dwarf.d:295]: normally I would but I want this to be identical with the LDC version

**Pull request merged**